### PR TITLE
torch bug fixed

### PR DESCRIPTION
**Pull request by Arrabonae**
this pull fixes the torch bug where the 'adaptive_sample' method keep sampling '=' which is token id 28. 
root cause was that initially the attn_varentropy (which is calculated in 'caclucalte_metrics') gives yo NaN instead of 0. as jax does, due to this, the 'adaptive_sample' method could not generate a reasonable sample, and keep falling back to token 28, that is '='. 

also added minor fixed around torch data types and such, including the proper handling of MacBook (apple silicon, cuda and cpu) for torch. 

torch version still 'easier' for users who use MacBooks as jax has no stable metal (mps) interface, only coda and cpu.
will plan to keep the torch version alive for a while for this exact reason.

### CODE REVIEW

**Comment by Arrabonae** [comment]
also, changed the poetry, to make the cuda 'optional' - same reason as i mentioned => for mac users

File: entropix/torch_main.py [hunk 1]
```diff
@@ -16,51 +18,71 @@
...
-        wavelen = 2 * math.pi / freq
-        if wavelen < high_freq_wavelen:
-            new_freqs.append(freq)
-        elif wavelen > low_freq_wavelen:
-            new_freqs.append(freq / scale_factor)
-        else:
-            assert low_freq_wavelen != high_freq_wavelen
-            smooth = (old_context_len / wavelen - low_freq_factor) / (
-                high_freq_factor - low_freq_factor
+from entropix.main import precompute_freqs_cis as jax_precompute_freqs_cis
```
Comment (trre123) [inline entropix/torch_main.py:21]: think this line can be removed since precompute_freqs_cis is defined below

Comment (Arrabonae) [reply entropix/torch_main.py:21]: yes, true. i have experimented with other approaches and this is a left over code from that

Comment (Arrabonae) [reply entropix/torch_main.py]: added the commit to this pr below, also removed he jax imports, same concepts: have experimented with other approaches but ended up leaving the torch implementation intact

**Comment by artus-LYTiQ** [comment]
I recommend to update the pyproject.toml to install torch on MacOS via PyPi, which is the recommended way. 

Change line 18 to 
`torch = [
    {version = "^2.4.1", source = "pypi", markers = "platform_system == 'Darwin'"},
    {version = "^2.4.1", source = "pytorch", extras = ["+cu124"], markers = "platform_system != 'Darwin'"}
]`

and then add a second "source" section before the build-system:
`[[tool.poetry.source]]
name = "pypi"
priority = "primary"`

**Review by 1976KALA — COMMENTED** [review]
Viewed

**Pull request merged**